File: python/arrays-and-hashing/contains-duplicate.py
class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        # Hash Map where the values will be of type int
        # Key: Number, Value: Count in Array
        counter = defaultdict(int)

        # O(n) loop through the numbers
        for i in nums:
            # Average case O(1) lookup
            if counter[i] == 1:
                return True

            # Increment the counter of that number i
            counter[i] += 1

        # If made it through the loop, can be assured there is no duplicate
        return False

File: c++/arrays-and-hashing/contains-duplicate.cc
class Solution {
public:
    bool containsDuplicate(vector<int>& nums) {
        std::unordered_map<int, bool> exists;
        for (int i = 0; i < nums.size(); ++i) {
            if (exists.contains(nums[i]) && exists.at(nums[i])) {
                return true;
            }
            exists[nums[i]] = true;
        }
        return false;
    }
};
